src/libraries/Common/src/Interop/Windows/Kernel32/Interop.REPARSE_DATA_BUFFER.cs

```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Kernel32
+    {
+        // https://docs.microsoft.com/windows-hardware/drivers/ifs/fsctl-get-reparse-point
+        public const int MAXIMUM_REPARSE_DATA_BUFFER_SIZE = 16 * 1024;
+
+        public const uint SYMLINK_FLAG_RELATIVE = 1;
+
+        // https://msdn.microsoft.com/library/windows/hardware/ff552012.aspx
+        // We don't need all the struct fields; omitting the rest.
+        [StructLayout(LayoutKind.Sequential)]
+        public unsafe struct REPARSE_DATA_BUFFER
+        {
+            public uint ReparseTag;
+            public ushort ReparseDataLength;
+            public ushort Reserved;
+            public SymbolicLinkReparseBuffer ReparseBufferSymbolicLink;
+
+            [StructLayout(LayoutKind.Sequential)]
+            public struct SymbolicLinkReparseBuffer
+            {
+                public ushort SubstituteNameOffset;
+                public ushort SubstituteNameLength;
+                public ushort PrintNameOffset;
+                public ushort PrintNameLength;
+                public uint Flags;
```

Comment:
The nested struct was added for clarity and easier manipulation.

Do you have any concerns around having a nested struct, as opposed to having it all flatly defined in the top struct?

Reply:
Only concern is that we will have to change this and define some structs if we will want to support other reparse points. (mount points and others). This is that we have in PowerShell code.

Reply:
I agree that the name of the nested union `SymbolicLinkReparseBuffer` should have a different, more generic name since it can represent any and all reparse points / tags.

I don't think flattening is necessary if C# allows it without having any negative side effects after passing it to the p/invoke.

Reply:
According to [the documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_reparse_data_buffer?redirectedfrom=MSDN), the C++ struct containing nested union structs can define the data for a symbolic link, for a mount point and the generic type:

```cpp
typedef struct _REPARSE_DATA_BUFFER {
  ULONG  ReparseTag;
  USHORT ReparseDataLength;
  USHORT Reserved;
  union {
    struct {
      USHORT SubstituteNameOffset;
      USHORT SubstituteNameLength;
      USHORT PrintNameOffset;
      USHORT PrintNameLength;
      ULONG  Flags;
      WCHAR  PathBuffer[1];
    } SymbolicLinkReparseBuffer;
    struct {
      USHORT SubstituteNameOffset;
      USHORT SubstituteNameLength;
      USHORT PrintNameOffset;
      USHORT PrintNameLength;
      WCHAR  PathBuffer[1];
    } MountPointReparseBuffer;
    struct {
      UCHAR DataBuffer[1];
    } GenericReparseBuffer;
  } DUMMYUNIONNAME;
} REPARSE_DATA_BUFFER, *PREPARSE_DATA_BUFFER;
```

Correct me if I'm wrong, but I think the code can be preserved as it is right now, and when the time comes to support junctions and others, we add the extra structs as needed.

Reply:
>  we add the extra structs as needed.

I mean we can not use the same name  REPARSE_DATA_BUFFER for all the structs.

Reply:
> I mean we can not use the same name REPARSE_DATA_BUFFER for all the structs.

Why not? My idea was to use REPARSE_DATA_BUFFER for all the pinvokes related to reparse points. Similar to what we have for symbolic links here:
https://github.com/dotnet/runtime/blob/55c1c67ad9653e82ca5d63f6bc47721321891af8/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs#L503

For mount points we would do `rdb.MountPointReparseBuffer.SubstituteNameOffset`.

Reply:
Using the same name for different purposes will be misleading. For example, there will be references with the same name in searches. This will make it difficult to understand and research the code.